satellite/satellitedb/overlaycache.go

```diff
@@ -33,112 +33,167 @@ type overlaycache struct {
+			finalArgs = append(finalArgs, tempNewNodeArgs...)
+			finalQuery = newNodeQuery + "  UNION ALL "
+		}
+		reputableNodeQuery, moreReputableNodeArgs := buildSelectionDistinct(ctx, criteria.ExcludedNetworks, reputableNodeCount, safeReputableNodeQuery, false)
+		tempReputableNodeArgs = append(tempReputableNodeArgs, moreReputableNodeArgs...)
+
+		finalQuery += reputableNodeQuery
+		finalArgs = append(finalArgs, tempReputableNodeArgs...)
+
+		rows, err := cache.db.Query(ctx, cache.db.Rebind(finalQuery), finalArgs...)
 		if err != nil {
-			return nil, err
+			return nil, Error.Wrap(err)
+		}
+		defer func() { err = errs.Combine(err, rows.Close()) }()
+
+		var moreNodes []*overlay.SelectedNode
+		for rows.Next() {
+			var isNew bool
+			var node overlay.SelectedNode
+			node.Address = &pb.NodeAddress{Transport: pb.NodeTransport_TCP_TLS_GRPC}
+			err = rows.Scan(&node.LastNet, &node.ID, &node.Address.Address, &node.LastIPPort, &isNew)
+			if err != nil {
+				return nil, err
+			}
+			// checking for last net collision among reputable and new nodes since we can't check within the query
+			if _, ok := receivedNodeNetworks[node.LastNet]; ok {
+				continue
+			}
+			if isNew {
+				receivedNewNodeCount++
+				receivedNodeNetworks[node.LastNet] = struct{}{}
+			}
+			moreNodes = append(moreNodes, &node)
 		}
 		for _, n := range moreNodes {
 			nodes = append(nodes, n)
 			criteria.ExcludedIDs = append(criteria.ExcludedIDs, n.ID)
 			criteria.ExcludedNetworks = append(criteria.ExcludedNetworks, n.LastNet)
 		}
-		if len(nodes) == count {
+		if len(nodes) == totalCount {
 			break
 		}
 	}
-
 	return nodes, nil
 }
 
-func (cache *overlaycache) SelectNewStorageNodes(ctx context.Context, count int, criteria *overlay.NodeCriteria) (nodes []*overlay.SelectedNode, err error) {
-	defer mon.Task()(&ctx)(&err)
+func buildConditions(ctx context.Context, criteria *overlay.NodeCriteria, isNewNodeQuery bool) (query string, args []interface{}, err error) {
 
 	nodeType := int(pb.NodeType_STORAGE)
 
-	safeQuery := `
-		WHERE disqualified IS NULL
-		AND suspended IS NULL
-		AND exit_initiated_at IS NULL
-		AND type = ?
-		AND free_disk >= ?
+	// initiated with reputable node conditions
+	reputationCondition := `
+	AND total_audit_count >= ?
+	AND total_uptime_count >= ?`
+
+	if isNewNodeQuery {
+		reputationCondition = `
 		AND (total_audit_count < ? OR total_uptime_count < ?)
-		AND last_contact_success > ?`
-	args := append(make([]interface{}, 0, 10),
-		nodeType, criteria.FreeDisk, criteria.AuditCount, criteria.UptimeCount, time.Now().Add(-criteria.OnlineWindow))
+		`
+	}
+
+	safeNodeQuery := `
+	WHERE disqualified IS NULL
+	AND suspended IS NULL
+	AND exit_initiated_at IS NULL
+	AND type = ?
+	AND free_disk >= ?
+	` + reputationCondition + `
+	AND last_contact_success > ?`
+	args = append(make([]interface{}, 0),
+		nodeType, criteria.FreeDisk, criteria.AuditCount,
+		criteria.UptimeCount, time.Now().Add(-criteria.OnlineWindow))
```

Comment:
It looks like, in the old code, criteria.UptimeCount always ended up being 0 when querying new nodes (it's not set in the NodeCriteria initialization at https://github.com/storj/storj/pull/3831/files#diff-5eccef64ca6f24b733402e9376a8f2fbL313 ). Is it correct to use a nonzero value here when `isNewNodeQuery` is true?

Reply:
good catch, I'll change that to be consistent with how it was before